src/zkp/pilog.rs

```diff
@@ -6,191 +6,311 @@
-            rho: rho.clone(),
-        }
+impl ProverSecret {
+    /// Collect prover secrets for proving [`PiLogProof`].
+    pub(crate) fn new(plaintext: BigNumber, nonce: Nonce) -> Self {
+        Self { plaintext, nonce }
     }
 }
 
-// Common input is: q, N0, C, X, g
-// Prover secrets are: (x, rho)
+/// Generates a challenge from a [`Transcript`] and the values generated in the
+/// proof.
+fn generate_challenge(
+    transcript: &mut Transcript,
+    common_input: &CommonInput,
+    plaintext_commit: &Commitment,
+    mask_encryption: &Ciphertext,
+    mask_dlog_commit: &CurvePoint,
+    mask_commit: &Commitment,
+) -> Result<BigNumber> {
+    transcript.append_message(b"Common input", &serialize!(&common_input)?);
+    transcript.append_message(
+        b"(plaintext commit, mask encryption, mask dlog commit, mask commit)",
+        &[
+            plaintext_commit.to_bytes(),
+            mask_encryption.to_bytes(),
+            serialize!(&mask_dlog_commit)?,
+            mask_commit.to_bytes(),
+        ]
+        .concat(),
+    );
+
+    // The challenge is sampled from `± q` (where `q` is the group order).
+    let challenge = plusminus_bn_random_from_transcript(transcript, &utils::k256_order());
+    Ok(challenge)
+}
+
 impl Proof for PiLogProof {
-    type CommonInput = PiLogInput;
-    type ProverSecret = PiLogSecret;
+    type CommonInput = CommonInput;
+    type ProverSecret = ProverSecret;
 
-    // N0: modulus, K: Paillier ciphertext
     #[cfg_attr(feature = "flame_it", flame("PiLogProof"))]
     fn prove<R: RngCore + CryptoRng>(
         input: &Self::CommonInput,
         secret: &Self::ProverSecret,
         transcript: &mut Transcript,
         rng: &mut R,
     ) -> Result<Self> {
-        // Sample alpha from plus/minus 2^{ELL + EPSILON}
-        let alpha = random_plusminus_by_size(rng, ELL + EPSILON);
-
-        let (S, mu) = input.setup_params.scheme().commit(&secret.x, ELL, rng);
-        let (A, r) = input.pk.encrypt(rng, &alpha)?;
-        let Y = CurvePoint(input.g.0 * utils::bn_to_scalar(&alpha)?);
-        let (D, gamma) = input
-            .setup_params
-            .scheme()
-            .commit(&alpha, ELL + EPSILON, rng);
-
-        transcript.append_message(b"CommonInput", &serialize!(&input)?);
-        transcript.append_message(
-            b"(S, A, Y, D)",
-            &[S.to_bytes(), A.to_bytes(), serialize!(&Y)?, D.to_bytes()].concat(),
-        );
-
-        // Verifier samples from e in +- q (where q is the group order)
-        let e = plusminus_bn_random_from_transcript(transcript, &input.q);
-
-        let z1 = &alpha + &e * &secret.x;
-        let z2 = input.pk.mask(&secret.rho, &r, &e);
-        let z3 = mu.mask(&gamma, &e);
-
-        let proof = Self {
-            alpha,
-            S,
-            A,
-            Y,
-            D,
-            e,
-            z1,
-            z2,
-            z3,
-        };
-
-        Ok(proof)
+        // The proof works as follows.
+        //
```

Comment:
This is really nice! Do you think it should be visible in the generated docs instead of only in the source code? I'm open either way, but it could be good to have it live at the module level.

Reply:
I was thinking just as a comment because my take was it's not too important to the caller how the underlying proof works, and is more for a reviewer of the code to validate it's being done the right way. I'm open to moving it to an externally-facing location though if you think that'd be useful!

Reply:
Yeah, that's fair. I know @indomitableSwan prefers to read all the generated docs first, but without the internal variables and types, this could be a bit confusing. Let's leave as-is for now, then.